### PR TITLE
Parallelise the Deconvolution

### PR DESCRIPTION
**Pull request by gordonkoehn**
This PR speeds up Lollipop's Deconvolution by Parallelization.

We implement a parallelization by `location`; thus, there is a speedup of roughly `1/n` given `n` locations, with little pre-and post-processing. 

The output of the code has been verified line by line with, i.e. `diff` to match the original code with sequential processing, i.e. `n-cores=1` [before safe thread random number generator seeding, see below.]

Considerations:
- Command Line Interface adjusted to remain backwards compatible, defaults to sequential processing
- Progression of Deconvolution also shown in Multiprocessing
- For Reproducibility even in Parallel Processing: Using thread-safe random number generation

### CODE REVIEW

File: lollipop/cli/deconvolute.py [hunk 1]
```diff
@@ -198,6 +413,22 @@ def deconvolute(
...
+            logging.warning(
+                "The number of cores is greater than the number of locations."
+            )
+            # adjust the number of cores to the number of locations
+            n_cores = len(loc) + 1
+            logging.warning(f"The number of cores has been adjusted to {n_cores}.")
+
+        # inform on the mode of computation
+        print("Available cores (parrallelized by locations): ", n_cores)
+
```
Comment (gordonkoehn) [inline lollipop/cli/deconvolute.py:431]: Primitive sanity checks for the number of cores assigned.

File: lollipop/cli/deconvolute.py [hunk 2]
```diff
@@ -309,7 +540,17 @@ def deconvolute(
...
+    # seed the random generator
+    # if locations_list is not or 1 we need one seed otherwise per location
+    if len(locations_list) == 1:
+        n_seeds = 1
+    else:
+        n_seeds = len(locations_list) + 1
+
+    seed_seq = np.random.SeedSequence(seed)
+    seeds = seed_seq.spawn(n_seeds)
```
Comment (gordonkoehn) [inline lollipop/cli/deconvolute.py:552]: [SeedSequences](https://numpy.org/doc/stable/reference/random/parallel.html) seems to be the recommended way to generate thread-safe random numbers for numpy.

File: lollipop/cli/deconvolute.py [hunk 3]
```diff
@@ -348,101 +589,72 @@ def deconvolute(
...
+    else:
+        # Create a pool of workers
+        with multiprocessing.Pool(processes=n_cores) as pool:
+            # Map the function to the arguments in parallel - choosing imap as objects are large instaed of pool.map
+            all_deconv = list(
+                tqdm(
+                    pool.imap(_deconvolute_bootstrap_wrapper, args_list),
+                    total=len(args_list),
+                    desc="All Locations",
+                    position=0,
```
Comment (gordonkoehn) [inline lollipop/cli/deconvolute.py:645]: The `python.multiprocessing` - I intentionally did choose `pool.imap` and not `pool.map`:
- we can get a nice progress bar as it retains data asynconously
- `pool.imap` is more memory-efficient than `pool.map` for large datasets, i.e. the dataframe with all locations is not copied until it is actually used – (not 100 % sure this is indeed correct)

File: lollipop/cli/deconvolute.py [hunk 4]
```diff
@@ -348,101 +589,72 @@ def deconvolute(
...
-                    res = t_kdec.fitted
-                    res["location"] = location
-                    all_deconv.append(res)
+    logging.info(f"all locations took {time.time() - start_time} seconds")
 
     print("post-process data")
+    # Flatten the results from pot. parrallel processing
+    all_deconv = [item for sublist in all_deconv for item in sublist]
+    # concatenate all the results
     deconv_df = pd.concat(all_deconv)
```
Comment (gordonkoehn) [inline lollipop/cli/deconvolute.py:656]: Some additional flattening is needed due to the extra `List` by parallel processing..

File: lollipop/preprocessors.py [hunk 5]
```diff
@@ -100,7 +102,7 @@ def general_preprocess(
         # self.df_tally = self
         self.df_tally = self.df_tally.replace(
             ["extra", "mut", "shared", "revert", "subset"], 1
-        ).infer_objects()
+        ).infer_objects(copy=False)
```
Comment (gordonkoehn) [inline lollipop/preprocessors.py:105]: This change is unrelated to the purpose of this PR. Yet, it fixes the current behaviour of the code to be compatible with future versions of pandas and hence silences a compatibility error that was continuously logged.

```FutureWarning: Downcasting behavior in `replace` is deprecated
and will be removed in a future version. To retain the old behavior,
explicitly call `result.infer_objects(copy=False)`.
To opt-in to the future behavior, set
`pd.set_option('future.no_silent_downcasting', True)`
s = s.replace(replace_dict)```

File: lollipop/cli/deconvolute.py [hunk 6]
```diff
@@ -26,6 +36,224 @@
...
+    confint_params : dict
+        The parameters for the confidence interval.
+    deconv_params : dict
+        The parameters for the deconvolution.
+    have_confint : bool
+        Whether to use a confidence interval.
+    confint_name : str
+        The name of the confidence interval.
+    namefield: str
+        The column to use as 'names' for the entries in the tally table.
```
Comment (gordonkoehn) [inline lollipop/cli/deconvolute.py:142]: integrated `namefield` - to clear the merge conflicts – already.

File: lollipop/cli/deconvolute.py [hunk 7]
```diff
@@ -26,6 +36,224 @@
...
+        if bootstrap > 1 and n_cores == 1
+        else [0]
+    ):
+        if bootstrap > 1:
+            # resample if we're doing bootstrapping
+            assert (
+                namefield in loc_df.columns
+            ), f"bootstrapping needs a column with names for the entries of the tally table, but no column '{namefield}' found. Use option '--namefield' to specify"
+            temp_dfb = ll.resample_mutations(
+                loc_df, loc_df[namefield].unique(), namefield
```
Comment (gordonkoehn) [inline lollipop/cli/deconvolute.py:170]: `namefield` again.

File: lollipop/cli/deconvolute.py [hunk 8]
```diff
@@ -302,7 +556,7 @@ def deconvolute(
             if maxdate:
                 assert (
                     mindate < maxdate
-                ), f"out of order dates: {mindate} >= {maxdate}. Please fix the content of {variants_date}"
+                ), f"out of order dates: {mindate} >= {maxdate}. Please fix the content of {variants_dates}"
```
Comment (gordonkoehn) [inline lollipop/cli/deconvolute.py:559]: I spotted one typo in the nested `ifs` by accident.

**Review by gordonkoehn — COMMENTED** [review]
I've implemented the improvements we discussed yesterday.
- one-off error in `zip`
- passing only `location` specific data frame to reduce the required memory.

Integrating your `namefield` changes has already made it ready to merge.

**Comment by gordonkoehn** [comment]
@DrYak It's ready to be merged. I already implemented and tested your `name field` changes. 

With the new `seeds` at `bootstraps=100`, we get a max variation up to `±0.004` eyeballing the `mean` for a few values.

**Comment by gordonkoehn** [comment]
To run this on Euler, I'd recommend `--mem-per-cpu=8000`  and `--cpus-per-task=8` for eight locations.

**Review by DrYak — APPROVED** [review]
:+1:

**Pull request merged**